shiny/render/_data_frame_utils/_tbl_data.py

```diff
@@ -43,356 +38,299 @@
-
-    return serialize_frame_pd(data)
-
-
-@serialize_frame.register
-def _(data: PlDataFrame) -> FrameJson:
-    import json
-
-    type_hints = list(map(serialize_dtype, data))
-    data_by_row = list(map(list, data.rows()))
-
-    # Shiny tag support
-    type_hints_type = [type_hint["type"] for type_hint in type_hints]
-    if "html" in type_hints_type:
-        session = require_active_session(None)
-
-        def wrap_shiny_html_with_session(x: TagNode):
-            return maybe_as_cell_html(x, session=session)
-
-        html_columns = [i for i, x in enumerate(type_hints_type) if x == "html"]
-
-        for html_column in html_columns:
-            for row in data_by_row:
-                row[html_column] = wrap_shiny_html_with_session(row[html_column])
-
-    data_val = json.loads(json.dumps(data_by_row, default=str))
+def serialize_frame(into_data: IntoDataFrame) -> FrameJson:
+
+    data = as_data_frame(into_data)
+
+    type_hints = [serialize_dtype(data[col_name]) for col_name in data.columns]
+
+    # TODO-future-barret; Swich serialization to "by column", rather than "by row"
+    # * This would allow for a single column to be serialized in a single operation
+    #   * This would allow for each column to capture the `"html"` type hint properly
+    #     for object and unknown columns. Currently, there is no way to determine which
+    #     cells are HTML-like during orjson serialization.
+    # * Even better, would be to move `orjson` serialization to the websocket
+    #   serialization.
+    #   * No need to serialize to JSON, then unserialize, then serialize again when
+    #     sending it to the client!
+    #   * The only serialization would occur during "send to browser"
+    #   * This approach would lose the ability to upgrade the column type hints to "html"
+    data_rows = data.rows(named=False)
+
+    session: Session | None = None
+
+    def default_orjson_serializer(val: Any) -> Jsonifiable:
+        nonlocal session
+        if ui_must_be_processed(val):
+            if session is None:
+                session = require_active_session(None)
+            return cast(JsonifiableDict, dict(as_cell_html(val, session=session)))
+
+        # All other values are serialized as strings
+        return str(val)
+
+    data_val = orjson.loads(
+        orjson.dumps(
+            data_rows,
+            default=default_orjson_serializer,
+            # option=(orjson.OPT_NAIVE_UTC),
+        )
+    )
 
     return {
-        # "index": list(range(len(data))),
         "columns": data.columns,
         "data": data_val,
         "typeHints": type_hints,
     }
 
 
 # subset_frame -------------------------------------------------------------------------
-def subset_frame__typed(
-    data: DataFrameLikeT, *, rows: RowsList = None, cols: ColsList = None
-) -> DataFrameLikeT:
-    return cast(DataFrameLikeT, subset_frame(data, rows=rows, cols=cols))
-
-
-@singledispatch
 def subset_frame(
```

Comment:
```
# subset_frame -------------------------------------------------------------------------
def subset_frame(
    data: DataFrameT,
    *,
    rows: RowsList = None,
    cols: ColsList = None,
) -> DataFrameT:
    """Return a subsetted DataFrame, based on row positions and column names.

    Note that when None is passed, all rows or columns get included.
    """

    # Note that this type signature assumes column names are strings things.
    # This is always true in Polars, but not in Pandas (e.g. a column name could be an
    # int, or even a tuple of ints)

    if rows is None and cols is None:
        return data
    row_idx = slice() if rows is None else list(rows)
    col_names = data.columns if cols is None else [data.columns[col] if isinstance(col, int) else col for col in cols]
    if len(col_names) == 0:
        return data[[], []]
    return data[row_idx, col_names]
```

Reply:
While I like the shorter code more, I ran into more issues that it was worth. 

It was difficult to support `slice()` or `slice(None)` as narwhals didn't appreciate it.

Keeping code as is